File: offer/15/JPhoebe.py
#         self.val = x
#         self.next = None

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def ReverseList(self, pHead):
        if pHead is None:
            return None
        pre = None
        next = None
        ## 交换法则
        while pHead is not None:
            next = pHead.next
            pHead.next = pre
            pre = pHead
            pHead = next
        return pre


value = ListNode({1, 2, 3, 4, 5})
value.next = ListNode({2, 3, 4, 5})
value.next.next = ListNode({3, 4, 5})
value.next.next.next = ListNode({4, 5})
value.next.next.next.next = ListNode({5})

print(Solution().ReverseList(value))

File: offer/55/JPhoebe.java

 public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}
public class Solution {

    public ListNode EntryNodeOfLoop(ListNode pHead) {
        if (pHead == null) {
            return null;
        }
        // 是否有环
        ListNode p1 = pHead, p2 = pHead;
        // 环最少有两个节点
        boolean isCircular = false;
        while (p2.next != null && p2.next.next != null) {
            p1 = p1.next;
            p2 = p2.next.next;
            if (p1 == p2) {
                isCircular = true;
                break;
            }
        }
        if (!isCircular) {
            return null;
        }

        // p1 和 p2 重合，任何一个点走一圈就是圆周长
        int length = 1;
        p2 = p2.next;
        while (p1 != p2) {
            ++length;
            p2 = p2.next;
        }

        // 从头开始，p2先走一圈，重逢的时候就是入口
        p1 = pHead;
        p2 = pHead;
        for (int i = 0; i < length; i++) {
            p2 = p2.next;
        }
        while (p1 != p2) {
            p1 = p1.next;
            p2 = p2.next;
        }
        return p1;
    }
}
